### PR TITLE
Static YML build pipeline

### PR DESCRIPTION
**Pull request by j0shuams**
This PR does two things

1) Converts our UI-based ADO pipeline to .yml file format. This makes version control much easier and makes our CI more extendable.

2) Adds Maestro publishing to our ADO pipeline. This will update a channel with CsWinRT releases, so our dependents can easily ingest new CsWinRT versions.

I've added all build/CI code to a new folder `build`, and Maestro helpers belong in folder `eng`, by convention. I include updates to our "repo structure" docs with this PR.

cc: @kythant (who was a tremendous help!)

### CODE REVIEW

File: eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml [hunk 1]
```diff
@@ -30,6 +30,39 @@ stages:
...
+   
+          @echo off
+   
+          rem Encode the build date/rev into a 16 bit value for resource versions
+          if "$(PrereleaseVersion)"=="" (
+            set RevisionBase=30000
+          ) else (
+            set RevisionBase=0
+          )
+          for /f "tokens=4,5 delims=." %%i in ("$(Build.BuildNumber)") do set BuildMetadata=%%i.%%j & set /a BuildRevision=%RevisionBase%+(((%%i/10000)-20)*366+((%%i)%%10000)/100*31+((%%i)%%100))*10+%%j
```
Comment (BenJKuhn) [inline eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:50]: I'm having a little trouble parsing what this expression is meant to produce. I'm not quite following the year & month calculations in build revision. My shell scripting is also a bit rusty, what does %%j represent?

Comment (j0shuams) [reply eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:50]: This is a bit arcane, this task gets us the version we're going to publish -- `1.4.2-prerelease.xxxxxx.x`
and this confusing bit calculates the `xxxxxx.x`, I believe. I think %%j might default to 0, but not sure. @Scottj1s  wrote this he might be able to elaborate.

**Comment by j0shuams** [comment]
@BenJKuhn @angelazhangmsft Ready for your review, confirmed that a test build gets published to maestro.

File: eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml [hunk 2]
```diff
@@ -0,0 +1,241 @@
+# Parameters:
+#   AssetNames and AssetVersions can be separated by ';' for publishing build with multiple assets
+#   The lengths when split by ';' must match between both of these parameters
+#     Example:
+#     AssetNames: "PackageName1;PackageName2"
+#     AssetVersions: "1.1;1.2"
+#   
+#   TriggerSubscription will publish the build to the default channel. 
+parameters:
```
Comment (BenJKuhn) [inline eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:9]: @kythant , are these folders supposed to be consistent across Maestro-driven repos? Any concerns about the customization here?

**Review by BenJKuhn — APPROVED** [review]

File: eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml (hunk 2 above)
Comment (kythant) [reply eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:9]: Talked with Joshua offline to revert these back to the original form and move the modifications outside in a new yml template.

File: build/AzurePipelineTemplates/CsWinRT-Pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# This file is used as basis for the ADO pipeline
+name: $(MajorVersion).$(MinorVersion).$(PatchVersion)$(PrereleaseVersion).$(date:yyMMdd)$(rev:.r)
+
+stages:
+- template: CsWinRT-BuildAndTest-Stage.yml
+
+- template: CsWinRT-PublishToNuget-Stage.yml
+
+# if we are doing a release build, publish the build to Maestro
+# - ${{ if eq( '$(PrereleaseVersion)', '') }}:
```
Comment (kythant) [inline build/AzurePipelineTemplates/CsWinRT-Pipeline.yml:10]: Does these needs to be uncommented?

File: eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml [hunk 4]
```diff
@@ -0,0 +1,168 @@
+parameters:
+  AssetName: ''
+  AssetVersion: ''
+  TriggerSubscription: true
+
```
Comment (kythant) [inline eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:5]: Please copy eng/common again from the main branch of ProjectReunionInternal as there has been updates.

File: build/AzurePipelineTemplates/CsWinRT-Pipeline.yml (hunk 3 above)
Comment (j0shuams) [reply build/AzurePipelineTemplates/CsWinRT-Pipeline.yml:10]: Yes this is just for testing and will be removed before merge

File: eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml (hunk 4 above)
Comment (j0shuams) [reply eng/common/AzurePipelineTemplates/Maestro-PublishBuildToMaestro-Steps.yml:14]: Thanks, done.

**Review by kythant — APPROVED** [review]

**Pull request merged**